keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/Cleaner.kt

```diff
@@ -0,0 +1,96 @@
+import org.springframework.transaction.annotation.Transactional
+import java.time.Clock
+
+/**
+ * A class used to orchestrate the deletion of delivery configs
+ */
+@Component
+class Cleaner(
+  private val deliveryConfigRepository: DeliveryConfigRepository,
+  private val artifactRepository: ArtifactRepository,
+  private val resourceRepository: ResourceRepository,
+  private val clock: Clock,
+  private val publisher: ApplicationEventPublisher
+) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  /**
+   * Deletes a delivery config and everything in it.
+   */
+  @Transactional(propagation = REQUIRED)
+  fun delete(deliveryConfigName: String): DeliveryConfig {
+    val deliveryConfig = deliveryConfigRepository.get(deliveryConfigName)
+
+    deliveryConfig.environments.forEach { environment ->
+      environment.resources.forEach { resource ->
+        // resources must be removed from the environment then deleted
+        deliveryConfigRepository.deleteResourceFromEnv(deliveryConfig.name, environment.name, resource.id)
+        delete(resource.id)
+      }
+      deliveryConfigRepository.deleteEnvironment(deliveryConfig.name, environment.name)
+    }
+
+    deliveryConfig.artifacts.forEach { artifact ->
+      artifactRepository.delete(artifact)
+    }
+
+    deliveryConfigRepository.delete(deliveryConfig.name)
+
+    return deliveryConfig
+  }
+
+  /**
+   * Removes artifacts, environments, and resources that were present in the [old]
+   * delivery config and are not present in the [new] delivery config
+   */
+  fun removeResources(old: DeliveryConfig, new: DeliveryConfig) {
+    val newResources = new.resources.map { it.id }
+    old.artifacts.forEach { artifact ->
+      if (artifact !in new.artifacts) {
+        log.debug("Updating config ${new.name}: removing artifact $artifact")
+        artifactRepository.delete(artifact)
+      }
+    }
+
+    old.environments
+      .forEach { environment ->
+        environment.resources.forEach { resource ->
+          if (resource.id !in newResources) {
+            log.debug("Updating config ${new.name}: removing resource ${resource.id} in environment ${environment.name}")
+            deliveryConfigRepository.deleteResourceFromEnv(
+              deliveryConfigName = old.name, environmentName = environment.name, resourceId = resource.id
+            )
+            delete(resource.id)
+          }
+        }
+      }
+
+    old.environments.forEach { environment ->
+      if (environment.name !in new.environments.map { it.name }) {
+        log.debug("Updating config ${new.name}: removing environment ${environment.name}")
+        deliveryConfigRepository.deleteEnvironment(new.name, environment.name)
+      }
+    }
+  }
+
+  fun delete(id: ResourceId): Resource<out ResourceSpec> =
+    resourceRepository
+      .get<ResourceSpec>(id)
+      .also {
+        resourceRepository.delete(id)
+        publisher.publishEvent(ResourceDeleted(it, clock))
```

Comment:
This is good. I think we're already missing events for operations on environments and delivery configs. Should we stretch this PR a bit and add a couple, at least for deletion events?

Reply:
Let's not stretch this PR - no reason to.